pkg/dataobj/sections/pointers/builder.go

```diff
@@ -0,0 +1,377 @@
+			// A stream
+			if b.pointers[i].StartTs.Equal(b.pointers[j].StartTs) {
+				return b.pointers[i].EndTs.Before(b.pointers[j].EndTs)
+			}
+			return b.pointers[i].StartTs.Before(b.pointers[j].StartTs)
+		}
+		// They're different, just make sure all the streams are separate from the columns
+		return b.pointers[i].StreamID < b.pointers[j].StreamID
+	})
+}
+
+func (b *Builder) encodeTo(enc *encoder) error {
+	// TODO(rfratto): handle one section becoming too large. This can happen when
+	// the number of columns is very wide. There are two approaches to handle
+	// this:
+	//
+	// 1. Split streams into multiple sections.
+	// 2. Move some columns into an aggregated column which holds multiple label
+	//    keys and values.
+
+	pathBuilder, err := dataset.NewColumnBuilder("path", dataset.BuilderOptions{
+		PageSizeHint: b.pageSize,
+		Value:        datasetmd.VALUE_TYPE_BYTE_ARRAY,
+		Encoding:     datasetmd.ENCODING_TYPE_PLAIN,
+		Compression:  datasetmd.COMPRESSION_TYPE_ZSTD,
+		Statistics: dataset.StatisticsOptions{
+			StoreRangeStats: true,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("creating path column: %w", err)
+	}
+	sectionBuilder, err := numberColumnBuilder(b.pageSize)
+	if err != nil {
+		return fmt.Errorf("creating section column: %w", err)
+	}
+
+	// Stream info
+	idBuilder, err := numberColumnBuilder(b.pageSize)
+	if err != nil {
+		return fmt.Errorf("creating ID column: %w", err)
+	}
+	streamIDRefBuilder, err := numberColumnBuilder(b.pageSize)
+	if err != nil {
+		return fmt.Errorf("creating stream ID in object column: %w", err)
+	}
+	minTimestampBuilder, err := numberColumnBuilder(b.pageSize)
+	if err != nil {
+		return fmt.Errorf("creating minimum timestamp column: %w", err)
+	}
+	maxTimestampBuilder, err := numberColumnBuilder(b.pageSize)
+	if err != nil {
+		return fmt.Errorf("creating maximum timestamp column: %w", err)
+	}
+	rowsCountBuilder, err := numberColumnBuilder(b.pageSize)
+	if err != nil {
+		return fmt.Errorf("creating rows column: %w", err)
+	}
+	uncompressedSizeBuilder, err := numberColumnBuilder(b.pageSize)
+	if err != nil {
+		return fmt.Errorf("creating uncompressed size column: %w", err)
+	}
+
+	// Column index info
+	columnNameBuilder, err := dataset.NewColumnBuilder("column_name", dataset.BuilderOptions{
+		PageSizeHint: b.pageSize,
+		Value:        datasetmd.VALUE_TYPE_BYTE_ARRAY,
+		Encoding:     datasetmd.ENCODING_TYPE_PLAIN,
+		Compression:  datasetmd.COMPRESSION_TYPE_ZSTD,
+		Statistics: dataset.StatisticsOptions{
+			StoreRangeStats: true,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("creating column name column: %w", err)
+	}
+
+	valuesBloomFilterBuilder, err := dataset.NewColumnBuilder("values_bloom_filter", dataset.BuilderOptions{
+		PageSizeHint: b.pageSize,
+		Value:        datasetmd.VALUE_TYPE_BYTE_ARRAY,
+		Encoding:     datasetmd.ENCODING_TYPE_PLAIN,
+		Compression:  datasetmd.COMPRESSION_TYPE_NONE, // TODO: is there a sensible compression algorithm for bloom filters?
```

Comment:
Any suggestions for this TODO?
I'm tempted to leave it as NONE for two reasons: Bloom bytes are likely fairly random so won't compress well, and so we don't have to spend CPU decompressing the blooms when reading them so they should be very fast to scan.
At some point I would like to try roaring bitmaps here though.